drivers/timer/arm_arch_timer.c

```diff
@@ -16,14 +16,21 @@
 #define MIN_DELAY	(1000)
 
 static struct k_spinlock lock;
-static volatile uint64_t last_cycle;
+static uint64_t last_cycle;
 
 static void arm_arch_timer_compare_isr(const void *arg)
 {
 	ARG_UNUSED(arg);
 
 	k_spinlock_key_t key = k_spin_lock(&lock);
 
+	/* Only the primary core is supposed to announce a tick */
```

Comment:
This will work, but is a little fragile in the face of unexpected latency on core 0.  The way this works on other SMP platforms is that all the interrupted CPUs independently compute (atomically!) the elapsed time from the last announced interrupt and announce it.  This will be zero on the second of two simultaneously-interrupted CPUs, and everything will work correctly.  And there will be less code, one fewer test on a hot path, and everything will be symmetric.

Reply:
Oh, actually belay that.  This won't work.  Timeslicing works by multiplexing the timer announce stream, which means that suppressing timer interrupts on the other cores means that timesliced threads on other CPUs won't be preempted as expected.

This is sort of a design wart.  At some point it would be better to rework timeslice expirations as a per-CPU timeout object and have the handler deliver the actual preemption via the scheduler IPI (or a more flexible derivative).

But at least for now it's required that every timeout be delivered to every CPU, even if the announce ticks parameter is zero.

Reply:
I pushed a new version that it works on tests but again not sure if it's the correct way to implement this because I'm not really familiar with the details of the kernel time keeping. Anyway, v2 just pushed.